File: appmonet-sdk/appmonet-sdk-mopub/src/main/java/com/monet/bidder/AppMonetNativeViewBinder.java
package com.monet.bidder;

import androidx.annotation.NonNull;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public class AppMonetNativeViewBinder {
    final int layoutId;
    final int mediaLayoutId;
    final int titleId;
    final int textId;
    final int callToActionId;
    final int iconId;

    @NonNull final Map<String, Integer> extras;

    private AppMonetNativeViewBinder(@NonNull Builder builder) {
        this.layoutId = builder.layoutId;
        this.mediaLayoutId = builder.mediaLayoutId;
        this.titleId = builder.titleId;
        this.textId = builder.textId;
        this.callToActionId = builder.callToActionId;
        this.iconId = builder.iconId;
        this.extras = builder.extras;
    }

    public static final class Builder {
        private final int layoutId;
        private int mediaLayoutId;
        private int titleId;
        private int textId;
        private int iconId;
        private int callToActionId;
        @NonNull
        private Map<String, Integer> extras = Collections.emptyMap();

        public Builder(int layoutId) {
            this.layoutId = layoutId;
            this.extras = new HashMap<>();
        }

        @NonNull
        public final Builder mediaLayoutId(int mediaLayoutId) {
            this.mediaLayoutId = mediaLayoutId;
            return this;
        }

        @NonNull
        public final Builder titleId(int titlteId) {
            this.titleId = titlteId;
            return this;
        }

        @NonNull
        public final Builder iconId(int iconId) {
            this.iconId = iconId;
            return this;
        }

        @NonNull
        public final Builder textId(int textId) {
            this.textId = textId;
            return this;
        }

        @NonNull
        public final Builder addExtras(final Map<String, Integer> resourceIds) {
            this.extras = new HashMap<>(resourceIds);
            return this;
        }

        @NonNull
        public final Builder addExtra(final String key, final int resourceId) {
            this.extras.put(key, resourceId);
            return this;
        }

        @NonNull
        public final Builder callToActionId(int callToActionId) {
            this.callToActionId = callToActionId;
            return this;
        }

        @NonNull
        public final AppMonetNativeViewBinder build() {
            return new AppMonetNativeViewBinder(this);
        }
    }
}

File: appmonet-sdk/appmonet-sdk-mopub/src/main/java/com/monet/bidder/AppMonetNativeViewBinder.kt
package com.monet.bidder

import java.util.HashMap

class AppMonetNativeViewBinder private constructor(builder: Builder) {
  val layoutId: Int
  val mediaLayoutId: Int
  val titleId: Int
  val textId: Int
  val callToActionId: Int
  val iconId: Int
  val extras: Map<String, Int>

  class Builder(internal val layoutId: Int) {
    internal var mediaLayoutId = 0
    internal var titleId = 0
    internal var textId = 0
    internal var iconId = 0
    internal var callToActionId = 0
    internal var extras = mutableMapOf<String, Int>()
    fun mediaLayoutId(mediaLayoutId: Int) = apply { this.mediaLayoutId = mediaLayoutId }

    fun titleId(titleId: Int) = apply { this.titleId = titleId }

    fun iconId(iconId: Int) = apply { this.iconId = iconId }

    fun textId(textId: Int) = apply { this.textId = textId }

    fun addExtras(resourceIds: Map<String, Int>?) = apply { this.extras = HashMap(resourceIds!!) }

    fun addExtra(
      key: String,
      resourceId: Int
    ) = apply { this.extras[key] = resourceId }

    fun callToActionId(callToActionId: Int) = apply {
      this.callToActionId = callToActionId
    }

    fun build() = AppMonetNativeViewBinder(this)
  }

  init {
    layoutId = builder.layoutId
    mediaLayoutId = builder.mediaLayoutId
    titleId = builder.titleId
    textId = builder.textId
    callToActionId = builder.callToActionId
    iconId = builder.iconId
    extras = builder.extras
  }
}